contracts/Settlement/SettlementLogic.sol

```diff
@@ -446,52 +528,148 @@ contract SettlementLogic is ISettlement, SettlementStorage {
             );
-            _feeLoanAssetByCollateral = sovrynSwapNetwork.rateByPath(
-                _path,
-                relayerFeeOnLoanAsset
+        } else if (order.collateralTokenSent > 0) {
+            relayerFee = _checkRelayerFee(
+                order.collateralTokenAddress,
+                order.collateralTokenSent,
+                order.collateralTokenSent,
+                false
             );
+        } else {
+            revert("Invalid loanTokenSent and collateralTokenSent");
         }
 
-        _checkRelayerFee(
-            relayerFee + _feeLoanAssetByCollateral,
-            order.collateralTokenAddress
+        actualCollateralAmount = order.collateralTokenSent.sub(relayerFee);
+        actualLoanTokenAmount = order.loanTokenSent.sub(
+            relayerFeeOnLoanAsset
         );
     }
 
-    function _checkRelayerFee(uint256 fee, address fromToken) internal view {
-        uint256 feeInRbtc = fee;
+    function _checkRelayerFee(
+        address fromToken,
+        uint256 orderSize,
+        uint256 amountToFill,
+        bool isSpot
+    ) internal view returns (uint256 relayerFee) {
+        uint256 estOrderFee = amountToFill.mul(relayerFeePercent).div(10**20);
+        uint256 minFeeAmount = isSpot ? minSwapOrderTxFee : minMarginOrderTxFee;
+        uint256 minFeeAmountInToken = minFeeAmount;
+
         if (fromToken != WRBTC_ADDRESS) {
             address[] memory path = sovrynSwapNetwork.conversionPath(
-                fromToken,
-                WRBTC_ADDRESS
+                WRBTC_ADDRESS,
+                fromToken
+            );
+            minFeeAmountInToken = sovrynSwapNetwork.rateByPath(
+                path,
+                minFeeAmount
             );
-            feeInRbtc = sovrynSwapNetwork.rateByPath(path, fee);
         }
-        require(
-            feeInRbtc > minFee,
-            "Order amount is too low to pay the relayer fee"
+
+        if (estOrderFee < minFeeAmountInToken) {
+            require(amountToFill > minFeeAmountInToken, "Order amount is too low to pay the relayer fee");
+            require(
+                orderSize > minFeeAmountInToken,
+                "Order amount is too low to pay the relayer fee"
+            );
+
+            relayerFee = minFeeAmountInToken;
+        } else {
+            uint256 fillAmountInRBtc = amountToFill;
+            uint256 minFillingAmount = isSpot
+                ? minSwapOrderSize
+                : minMarginOrderSize;
+            if (fromToken != WRBTC_ADDRESS) {
+                address[] memory pathToRbtc = sovrynSwapNetwork.conversionPath(
+                    fromToken,
+                    WRBTC_ADDRESS
+                );
+                fillAmountInRBtc = sovrynSwapNetwork.rateByPath(
+                    pathToRbtc,
+                    amountToFill
+                );
+            }
+
+            require(
+                fillAmountInRBtc >= minFillingAmount,
+                "Filling amount is too low"
```

Comment:
I think this block is checking the min order size and not min transaction fees, as such `amountToFill>= minFeeAmountInToken` check probably won't work. Keeping it open for further discussion.

Reply:
this function lacks documentation. It took me 15 minutes to understand what was happening here, akthough it's just determining the relayer fee and checking amounts. 

line 599 - 604 -> orderSize > amountToFill, so the second requirement check doesn't make sense. 

The first price conversion in L 593 should be done using the oracle instead of the AMM. querying the price from the AMM is expensive.

The else block from line 607 doesn't make much sense to me.If the estOrderFee > minFee, why would we need to check that the amount which is filles would be big enough? If the amount to fill is big enough so that 0.2% are enough to cover the min fee, the fill amount has to be high enough. I think these checks are not required, but let me know if i'm wrong.

Reply:
Done: https://github.com/DistributedCollective/LimitOrder/commit/51f795b399242e4ba127349724a04718c9312970 and https://github.com/DistributedCollective/LimitOrder/commit/75f297450aa49540d43fea131cef5b9fa7686a47